File: day_4/Python/camp_cleanup/main.py
def part_1(input_str):
    pairs = input_str.split("\n")

    total = 0
    for pair in pairs:
        left, right = [[int(s) for s in i] for i in [j.split("-") for j in pair.split(",")]]

        if left[0] <= right[0] and left[1] >= right[1]:
            total += 1

        elif right[0] <= left[0] and right[1] >= left[1]:
            total += 1

    return total


def part_2(input_str):
    pairs = input_str.split("\n")

    total = 0
    for pair in pairs:
        left, right = [[int(s) for s in i] for i in [j.split("-") for j in pair.split(",")]]

        if not (left[1] < right[0] or left[0] > right[1]):
            total += 1

    return total


def main(input_str):
    print(part_1(input_str))
    print(part_2(input_str))

if __name__ == "__main__":
    main(input_str)

File: day_4/Rust/camp_cleanup/src/main.rs
fn part_1(input_str: &str) -> i64 {
    let mut total = 0;
    let pairs = input_str.trim().split("\n");

    for pair in pairs {
        let pair = pair.trim().split(",").collect::<Vec<&str>>();

        let left = pair[0]
            .split("-")
            .map(|x| x.parse::<i32>().unwrap())
            .collect::<Vec<i32>>();

        let right = pair[1]
            .split("-")
            .map(|x| x.parse::<i32>().unwrap())
            .collect::<Vec<i32>>();

        if left[0] <= right[0] && left[1] >= right[1] {
            total += 1
        }

        else if right[0] <= left[0] && right[1] >= left[1] {
            total += 1
        }
    }
    total
}

fn part_2(input_str: &str) -> i64 {
    let mut total = 0;
    let pairs = input_str.trim().split("\n");

    for pair in pairs {
        let pair = pair.trim().split(",").collect::<Vec<&str>>();

        let left = pair[0]
            .split("-")
            .map(|x| x.parse::<i32>().unwrap())
            .collect::<Vec<i32>>();

        let right = pair[1]
            .split("-")
            .map(|x| x.parse::<i32>().unwrap())
            .collect::<Vec<i32>>();

        if !(left[1] < right[0] || left[0] > right[1]) {
            total += 1
        }
    }
    total
}

fn main() {
    let input_str = String::from("");

    println!("{}", part_1(&input_str));
    println!("{}", part_2(&input_str));
}
